### PR TITLE
Enable gateways to resolve hostnames to IPv4 addresses

### PR DESCRIPTION
**Pull request by freddygv**
This PR enables Terminating Gateways and Mesh Gateways to resolve hostnames.

The DNS resolution will be handled by Envoy and defaults to [LOGICAL_DNS](https://www.envoyproxy.io/docs/envoy/v1.14.1/intro/arch_overview/upstream/service_discovery#logical-dns). This discovery type can be overridden on a per-gateway basis with the `envoy_dns_discovery_type` [Gateway Option](https://github.com/hashicorp/consul/blob/21ff48eca797d0926f9a980e3fb95c8f95e2f7ef/website/pages/docs/connect/proxies/envoy.mdx#gateway-options).

If a service contains an instance with a hostname as an address we set the Envoy cluster to use DNS as the discovery type rather than EDS. Since both mesh gateways and terminating gateways route to clusters using SNI, whenever there is a mix of hostnames and IP addresses associated with a service we use the hostname + CDS rather than the IPs + EDS.

Note that we detect hostnames by attempting to parse the service instance's address as an IP. If it is not a valid IP we assume it is a hostname.

### CODE REVIEW

File: agent/proxycfg/snapshot.go [hunk 1]
```diff
@@ -100,6 +100,10 @@ type configSnapshotTerminatingGateway struct {
 	// between the gateway and a service. TLS configuration stored here is
 	// used for TLS origination from the gateway to the linked service.
 	GatewayServices map[structs.ServiceID]structs.GatewayService
+
+	// HostnameDatacenters is a map of service id to service instances with a hostname as the address.
```
Comment (banks) [inline agent/proxycfg/snapshot.go:104]:
```suggestion
	// HostnameServices is a map of service id to service instances with a hostname as the address.
```

File: agent/proxycfg/state.go [hunk 2]
```diff
@@ -1518,3 +1541,35 @@ func (s *state) Changed(ns *structs.NodeService, token string) bool {
 		!reflect.DeepEqual(s.proxyCfg, proxyCfg) ||
 		s.token != token
 }
+
+// When endpoints come in for mesh gateways or external services we check to see if any instances have a hostname their address.
```
Comment (banks) [inline agent/proxycfg/state.go:1545]: nit: I still prefer to format method comments according to go conventions even for private methods. This comment is useful context but perhaps we could reword the first line to match conventional way of describing the intent of a method?

```suggestion
// hostnameEndpoints returns all CheckServiceNodes that have hostnames instead of IPs as the address.
//
```

File: agent/xds/clusters.go [hunk 3]
```diff
@@ -133,32 +133,56 @@ func (s *Server) clustersFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsho
...
+				Name: cfgSnap.ServerSNIFn(dc, ""),
+			}
+			err := s.populateGatewayCluster(cfgSnap, cluster, 0, useEDS)
 			if err != nil {
 				return nil, err
 			}
+			if !useEDS {
+				injectHostnameEndpoints(cluster, cfgSnap.MeshGateway.HostnameDatacenters[dc], dc != cfgSnap.Datacenter, false)
+			}
+
```
Comment (banks) [inline agent/xds/clusters.go:176]: This block and the one above it seem to be repeated. Would it be possible to move all of this logic inside `makeGatewayCluster` and kept the old code here? We already pass `cfgSnap` so it should have all the info it needs to make do the same logic unless I'm missing something.

File: agent/xds/testdata/endpoints/terminating-gateway-default-service-subset.golden [hunk 4]
```diff
@@ -1,40 +1,6 @@
...
-                  }
-                }
-              },
-              "healthStatus": "HEALTHY",
-              "loadBalancingWeight": 1
-            }
-          ]
-        }
-      ]
-    },
```
Comment (banks) [inline agent/xds/testdata/endpoints/terminating-gateway-default-service-subset.golden:37]: I didn't follow the test fixture change in enough detail but this is expected right? That existing tests that didn't use hostnames should have their results change?

File: agent/xds/testdata/listeners/mesh-gateway-custom-addresses.golden [hunk 5]
```diff
@@ -168,6 +216,22 @@
...
+          "filters": [
+            {
+              "name": "envoy.tcp_proxy",
+              "config": {
+                  "cluster": "dc4.internal.11111111-2222-3333-4444-555555555555.consul",
+                  "stat_prefix": "mesh_gateway_remote_foo_dc4_tcp"
+                }
+            }
+          ]
+        },
```
Comment (banks) [inline agent/xds/testdata/listeners/mesh-gateway-custom-addresses.golden:234]: Why would these change for mesh gateways? This doesn't seem related to the issue being fixed here. Is this an unintended side effect of the logic change?

File: website/pages/docs/connect/proxies/envoy.mdx [hunk 6]
```diff
@@ -319,6 +319,10 @@ will continue to be supported.
 - `envoy_gateway_no_default_bind` - Prevents binding to the default address
   of the gateway service. This should be used with one of the other options
   to configure the gateway's bind addresses.
+  
+- `envoy_dns_discovery_type` - Determines how Envoy will resolve hostnames. Defaults to `LOGICAL_DNS`.
```
Comment (banks) [inline website/pages/docs/connect/proxies/envoy.mdx:323]: Should we clarify that this is only used for external services via a terminating gateway right now?

File: agent/proxycfg/state.go (hunk 2 above)
Comment (freddygv) [reply agent/proxycfg/state.go:1545]: Sounds good to me. Fixed.

File: agent/xds/testdata/endpoints/terminating-gateway-default-service-subset.golden (hunk 4 above)
Comment (freddygv) [reply agent/xds/testdata/endpoints/terminating-gateway-default-service-subset.golden:37]: Yep, this is expected. 

I updated the API service so that it has a mix of hostnames/IPs rather than just IPs here:
https://github.com/hashicorp/consul/pull/7999/files#diff-f8966ffd7dc2c27418ed8181bcedeecdR1620

Alternatively I could add a third service rather than modify one of the existing ones.

File: agent/xds/testdata/listeners/mesh-gateway-custom-addresses.golden (hunk 5 above)
Comment (freddygv) [reply agent/xds/testdata/listeners/mesh-gateway-custom-addresses.golden:234]: I added dc4 to the default test snapshot for mesh gateways here:
https://github.com/hashicorp/consul/pull/7999/files#diff-f8966ffd7dc2c27418ed8181bcedeecdR1390

Now in all mesh gateway tests we create an additional `filterChainMatch` for traffic to dc4.

File: website/pages/docs/connect/proxies/envoy.mdx (hunk 6 above)
Comment (freddygv) [reply website/pages/docs/connect/proxies/envoy.mdx:323]: This applies to both mesh gateways and terminating gateways, so I added a note about that.

File: agent/xds/clusters.go (hunk 3 above)
Comment (freddygv) [reply agent/xds/clusters.go:176]: Refactored this

File: agent/xds/clusters.go [hunk 7]
```diff
@@ -133,32 +133,56 @@ func (s *Server) clustersFromSnapshotMeshGateway(cfgSnap *proxycfg.ConfigSnapsho
...
+		}
+
 		clusters = append(clusters, cluster)
 	}
 
 	if cfgSnap.ServiceMeta[structs.MetaWANFederationKey] == "1" && cfgSnap.ServerSNIFn != nil {
 		// Add all of the remote wildcard datacenter mappings for servers.
 		for _, dc := range datacenters {
-			clusterName := cfgSnap.ServerSNIFn(dc, "")
+			if dc == cfgSnap.Datacenter {
```
Comment (rboyer) [inline agent/xds/clusters.go:158]: This section is about setting up the wildcard clusters for getting OUT to another datacenter `(dc != cfgSnap.Datacenter)` and *also* one for getting IN from another datacenter `(dc == cfgSnap.Datacenter)`.

You should not omit the local dc here, but you still should handle it differently now. When `dc == cfgSnap.Datacenter` it should behave like before, since it's definitely not going to be passing hostnames here, since we're using our real local catalog entries which are going to have full IP addresses for our own servers.

Comment (freddygv) [reply agent/xds/clusters.go:153]: Ah, yep that fixed it.

File: agent/xds/clusters.go [hunk 8]
```diff
@@ -547,43 +552,70 @@ func makeClusterFromUserConfig(configJSON string) (*envoy.Cluster, error) {
...
-	return s.makeGatewayClusterWithConnectTimeout(cfgSnap, clusterName, 0)
-}
-
-// makeGatewayClusterWithConnectTimeout initializes a gateway cluster
-// with the specified connect timeout. If the timeout is 0, the connect timeout
-// defaults to use the configured gateway timeout.
-func (s *Server) makeGatewayClusterWithConnectTimeout(cfgSnap *proxycfg.ConfigSnapshot,
-	clusterName string, connectTimeout time.Duration) (*envoy.Cluster, error) {
+func (s *Server) makeGatewayCluster(snap *proxycfg.ConfigSnapshot,
+	name string, hostnameEndpoints structs.CheckServiceNodes, isRemote, onlyPassing bool, connectTimeout time.Duration) *envoy.Cluster {
```
Comment (rboyer) [inline agent/xds/clusters.go:556]: This function has a bunch of args now. It _might_ be worth it for clarity to use an options struct, especially since there are now two boolean args which are easy to mix up.

File: agent/xds/clusters.go [hunk 9]
```diff
@@ -547,43 +552,70 @@ func makeClusterFromUserConfig(configJSON string) (*envoy.Cluster, error) {
...
+	cluster.DnsLookupFamily = envoy.Cluster_V4_ONLY
+
+	discoveryType := envoy.Cluster_Type{Type: envoy.Cluster_LOGICAL_DNS}
+	if cfg.DNSDiscoveryType == "strict_dns" {
+		discoveryType.Type = envoy.Cluster_STRICT_DNS
 	}
+	cluster.ClusterDiscoveryType = &discoveryType
 
-	return &cluster, nil
+	var endpoints []envoyendpoint.LbEndpoint
```
Comment (rboyer) [inline agent/xds/clusters.go:606]:
```suggestion
	endpoints := make([]envoyendpoint.LbEndpoint, 0, len(hostnameEndpoints))
```
This is not likely to matter in practice, but it doesn't hurt.
